File: jdbc-template/src/main/java/io/thomastodon/spring/data/examples/house/HouseDtoToHouseTranslator.java
package io.thomastodon.spring.data.examples.house;

import static java.util.stream.Collectors.toList;

public class HouseDtoToHouseTranslator implements Translator<HouseDto, House> {

    @Override
    public House translate(HouseDto source) {

        return new House(
            source.getUuid(),
            source.getRooms().stream().map(roomDto -> new Room(
                roomDto.getUuid(),
                roomDto.getChairs().stream().map(chairDto -> new Chair(
                    chairDto.getUuid(),
                    chairDto.getLegs().stream().map(legDto -> new Leg(legDto.getUuid())).collect(toList())
                )).collect(toList())
            )).collect(toList())
        );
    }
}

File: jdbc-template/src/main/java/io/thomastodon/spring/data/examples/house/HouseDtoToHouseTranslator.kt
package io.thomastodon.spring.data.examples.house

import java.util.*

class HouseDtoToHouseTranslator {

    fun translate(houseUuid: UUID, houseDto: HouseDto): House {

        return House(
            houseUuid,
            houseDto.rooms.map { (roomUuid, roomDto) ->
                Room(
                    roomUuid,
                    roomDto.chairs.map { (chairUuid, chairDto) ->
                        Chair(
                            chairUuid,
                            chairDto.legs.map { (legUuid, legDto) -> Leg(legUuid) }
                        )
                    }
                )
            }
        )
    }
}
